File: src/onion/compiler/MultiTable.java
package onion.compiler;

import java.util.*;

/**
 * Created by IntelliJ IDEA.
 * User: Mizushima
 * Date: 2010/12/07
 * Time: 2:48:39
 * To change this template use File | Settings | File Templates.
 */
public class MultiTable<E extends Named> implements Iterable<E> {
  private final Map<String, List<E>> mapping;
  public MultiTable() {
    this.mapping =  new HashMap<String, List<E>>();
  }
  public boolean add(E entry) {
    List<E> value = mapping.get(entry.name());
    if(value == null) {
      value = new ArrayList<E>();
      value.add(entry);
      mapping.put(entry.name(), value);
      return false;
    }else {
      value.add(entry);
      return true;
    }
  }
  public List<E> get(String key) {
    return mapping.get(key);
  }
  public List<E> values() {
    List<E> list = new ArrayList<E>();
    for(List<E> value:mapping.values()) list.addAll(value);
    return list;
  }
  public Iterator<E> iterator() {
    return values().iterator();
  }
}

File: src/main/scala/onion/compiler/MultiTable.scala
package onion.compiler

import scala.collection.Iterable
import scala.collection.Iterator
import scala.collection.mutable

class MultiTable[E <: Named] extends Iterable[E] {
  // Entries per name are kept as an immutable List, appended on add (adds are rare and the
  // lists are short -- they hold one name's overloads) so that `get`, which method
  // resolution calls for every class up a receiver's hierarchy on every call, returns the
  // list itself. The previous shape allocated on every lookup: a hit copied the buffer with
  // toList, and a *miss* inserted an empty buffer into the map -- a read that grew the
  // table for every name ever asked of every class, `size` on `Object` included.
  private[this] final val mapping = new mutable.HashMap[String, List[E]]

  // `values` is asked for on every method lookup over a user class; it is rebuilt only
  // after an `add`.
  private[this] var cachedValues: List[E] = null

  def add(entry: E): Boolean = {
    cachedValues = null
    mapping.get(entry.name) match {
      case Some(v) =>
        mapping(entry.name) = v :+ entry
        true
      case None =>
        mapping(entry.name) = entry :: Nil
        false
    }
  }

  def get(key: String): Seq[E] = mapping.getOrElse(key, Nil)

  def values: Seq[E] = {
    if (cachedValues == null) cachedValues = mapping.values.toList.flatten
    cachedValues
  }
  def iterator: Iterator[E] = values.iterator
}
